### DISCUSSION TITLE
issues with default include path for libev

### DISCUSSION THREAD

**Discussion by ndlarsen** [Homebrew/discussions#3431] [Everyday usage] [answered]
I am a little unsure where to place this question as this is related to both `libev` and `bjoern` (python library installed in the project via pip/poetry). My initial thought is that this is an issue related to the libev formulae but correct me if I am wrong.

The issue I and several of my colleagues are experiencing is that the library bjorn fails to locate `ev.h`. This makes sense as bjoern attempts to locate the file in a set of predefined and common locations and `/usr/local/Cellar/libev/4.33/include/` is naturally not one of them nor should it be.
Also, I do not have the folder `/opt/homebrew` which most people seem to have. I installed homebrew as instructed on https://brew.sh/

Unless something has changed in later versions of homebrew, I think something might not be entirely right and any thoughts on this are welcome.

Besides manually creating a symlink, which I consider to be a workaround and not a solution, or a pull request to bjoern, any ideas on this?

```console
$ sw_vers
ProductName:	macOS
ProductVersion:	12.4
BuildVersion:	21F79
```

```console
$ brew --version
Homebrew 3.5.4
Homebrew/homebrew-core (git revision dd54c13d569; last commit 2022-07-15)
Homebrew/homebrew-cask (git revision b0d9a56b46; last commit 2022-07-15)
```

I have installed `libev` via brew as this is required by some of the dependencies used in a project at work.

```console
$ brew install -v libev
...
==> Verifying checksum for '13922205a2d11a1217db2ff410702ec422fac0f53496b13ff1966085c20d9787--libev--4.33.monterey.bottle.tar.gz'
==> Pouring libev--4.33.monterey.bottle.tar.gz
tar --extract --no-same-owner --file /Users/username/Library/Caches/Homebrew/downloads/13922205a2d11a1217db2ff410702ec422fac0f53496b13ff1966085c20d9787--libev--4.33.monterey.bottle.tar.gz --directory /private/tmp/d20220715-33190-dkya84
cp -pR /private/tmp/d20220715-33190-dkya84/libev/. /usr/local/Cellar/libev
chmod -Rf +w /private/tmp/d20220715-33190-dkya84
==> Finishing up
ln -s ../Cellar/libev/4.33/include/ev++.h ev++.h
ln -s ../Cellar/libev/4.33/include/ev.h ev.h
ln -s ../../../Cellar/libev/4.33/share/man/man3/ev.3 ev.3
ln -s ../Cellar/libev/4.33/lib/libev.4.dylib libev.4.dylib
ln -s ../Cellar/libev/4.33/lib/libev.a libev.a
ln -s ../Cellar/libev/4.33/lib/libev.dylib libev.dylib
==> Summary
🍺  /usr/local/Cellar/libev/4.33: 12 files, 483.4KB
==> Running `brew cleanup libev`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

```console
$ brew list -v libev
/usr/local/Cellar/libev/4.33/INSTALL_RECEIPT.json
/usr/local/Cellar/libev/4.33/LICENSE
/usr/local/Cellar/libev/4.33/.brew/libev.rb
/usr/local/Cellar/libev/4.33/Changes
/usr/local/Cellar/libev/4.33/include/ev++.h
/usr/local/Cellar/libev/4.33/include/ev.h
/usr/local/Cellar/libev/4.33/README
/usr/local/Cellar/libev/4.33/TODO
/usr/local/Cellar/libev/4.33/lib/libev.a
/usr/local/Cellar/libev/4.33/lib/libev.dylib
/usr/local/Cellar/libev/4.33/lib/libev.4.dylib
/usr/local/Cellar/libev/4.33/share/man/man3/ev.3
```

```console
$ ls /usr/local/lib | grep ev
libev.4.dylib
libev.a
libev.dylib
libevent-2.1.7.dylib
libevent.a
libevent.dylib
libevent_core-2.1.7.dylib
libevent_core.a
libevent_core.dylib
libevent_extra-2.1.7.dylib
libevent_extra.a
libevent_extra.dylib
libevent_openssl-2.1.7.dylib
libevent_openssl.a
libevent_openssl.dylib
libevent_pthreads-2.1.7.dylib
libevent_pthreads.a
libevent_pthreads.dylib
```

Regards

**Comment by carlocab** [comment, marked as answer]
Is there no `ev.h` in `/usr/local/include`?


What is the output of `brew config`, `brew doctor` and `brew info libev`?

**Reply by ndlarsen** [reply]
I appreciate your reply.

> Is there no `ev.h` in `/usr/local/include`?

There is. `/usr/local/include` is not part of bjoern's default `include_dirs`, though.

```console
$ ls /usr/local/include | grep ev
ev++.h
ev.h
evdns.h
event.h
event2
evhttp.h
evrpc.h
evutil.h
unbound-event.h
```

> What is the output of `brew config`, `brew doctor` and `brew info libev`?

```console
$ brew config
HOMEBREW_VERSION: 3.5.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 88a5fac96752f17280e64f60926f58dff45990e0
Last commit: 10 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4ad3f8dbedd51a75c800157423092792f9369a7a
Core tap last commit: 9 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 13.1.6 build 1316
Git: 2.36.1 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.4-x86_64
CLT: 13.4.0.0.1.1651278267
Xcode: N/A
```
```console
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  poetry

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
```

```console
$ ﻿﻿brew info libev
libev: stable 4.33 (bottled)
Asynchronous event library
http://software.schmorp.de/pkg/libev.html
/usr/local/Cellar/libev/4.33 (12 files, 483.4KB) *
  Poured from bottle on 2022-07-15 at 10:49:54
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libev.rb
==> Analytics
install: 5,277 (30 days), 17,101 (90 days), 665,596 (365 days)
install-on-request: 646 (30 days), 1,903 (90 days), 9,756 (365 days)
build-error: 0 (30 days)
```

Does all of the above look correct? If so, I guess a PR against bjoern would be the next logical step.
Also, did something change in later versions of Homebrew causing `/opt/homebrew/` to no longer be default?

**Reply by carlocab** [reply]
That looks mostly in order, though it would be good to fix the warnings from `brew doctor`. Have you tried doing
```
pip3 install --global-option=build_ext --global-option="-I/usr/local/include" --global-option="-L/usr/local/lib" bjoern
```
which is roughly the suggestion in https://github.com/jonashaag/bjoern/issues/229#issuecomment-1146548084 (modified slightly)?

**Reply by carlocab** [reply]
> Also, did something change in later versions of Homebrew causing `/opt/homebrew/` to no longer be default?

Not exactly. For the longest time, `/usr/local` was the default prefix for Homebrew. Then, Linux support was added, where the default prefix is `/home/linuxbrew/.linuxbrew`. When Apple introduced M1/ARM64 Macs, the default prefix of `/opt/homebrew` was chosen for those.

**Reply by ndlarsen** [reply]
> That looks mostly in order, though it would be good to fix the warnings from `brew doctor`. Have you tried doing
> 
> ```
> pip3 install --global-option=build_ext --global-option="-I/usr/local/include" --global-option="-L/usr/local/lib" bjoern
> ```
> 
> which is roughly the suggestion in [jonashaag/bjoern#229 (comment)](https://github.com/jonashaag/bjoern/issues/229#issuecomment-1146548084) (modified slightly)?

I did see this workaround mentioned on their issue tracker but no. This would probably solve problem for projects where pip is used directly but we are currently using poetry across the board. I am also looking for a long term solution which does not require us to jump through hoops when e.g. onboarding new members. PR it is.

Thanks again.

**Reply by ndlarsen** [reply]
> > Also, did something change in later versions of Homebrew causing `/opt/homebrew/` to no longer be default?
> 
> Not exactly. For the longest time, `/usr/local` was the default prefix for Homebrew. Then, Linux support was added, where the default prefix is `/home/linuxbrew/.linuxbrew`. When Apple introduced M1/ARM64 Macs, the default prefix of `/opt/homebrew` was chosen for those.

Right, being on a 2019 pro (or so I am told) is makes sense that I do not have this directory, then.

**Reply by ndlarsen** [reply]
I appreciate your replies and will mark this topic as answered.

**Reply by carlocab** [reply]
Ah, yes. Poetry does seem not to support the extra flags you want to pass yet: https://github.com/python-poetry/poetry/issues/845

Something that _might_ help (since I suspect `poetry` just invokes `setup.py` in the background) is to do
```sh
export CPPFLAGS=-I/usr/local/include
export CFLAGS=-I/usr/local/include
export LDFLAGS=-L/usr/local/lib
```

**Reply by ndlarsen** [reply]
> Ah, yes. Poetry does seem not to support the extra flags you want to pass yet: [python-poetry/poetry#845](https://github.com/python-poetry/poetry/issues/845)
> 
> Something that _might_ help (since I suspect `poetry` just invokes `setup.py` in the background) is to do
> 
> ```shell
> export CPPFLAGS=-I/usr/local/include
> export CFLAGS=-I/usr/local/include
> export LDFLAGS=-L/usr/local/lib
> ```

I can confirm this works as a temporary workaround.

**Reply by ndlarsen** [reply]
Created a PR for bjoern [here](https://github.com/jonashaag/bjoern/pull/235).
